src/ClientGenerator/CodeGenerator.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Extensions.Logging;
+using Orleans.Logging;
+#if !NET461
+using System.Runtime.Loader;
+using Orleans.Runtime.Configuration;
+using Orleans.CodeGenerator;
+using Orleans.Serialization;
+#endif
+
+namespace Orleans.CodeGeneration
+{
+    public class CodeGenerator
+    {
+        private static readonly int[] SuppressCompilerWarnings =
+        {
+            162, // CS0162 - Unreachable code detected.
+            219, // CS0219 - The variable 'V' is assigned but its value is never used.
+            414, // CS0414 - The private field 'F' is assigned but its value is never used.
+            649, // CS0649 - Field 'F' is never assigned to, and will always have its default value.
+            693, // CS0693 - Type parameter 'type parameter' has the same name as the type parameter from outer type 'T'
+            1591, // CS1591 - Missing XML comment for publicly visible type or member 'Type_or_Member'
+            1998 // CS1998 - This async method lacks 'await' operators and will run synchronously
+        };
+
+        private readonly CodeGenOptions options;
+        private readonly Action<string> log;
+
+        public CodeGenerator(CodeGenOptions options, Action<string> log)
+        {
+            this.options = options;
+            this.log = log;
+        }
+
+        public bool GenerateCode()
+        {
+            var outputFileName = this.options.OutputFileName;
+
+            // Create directory for output file if it does not exist
+            var outputFileDirectory = Path.GetDirectoryName(outputFileName);
+
+            if (!string.IsNullOrEmpty(outputFileDirectory) && !Directory.Exists(outputFileDirectory))
+            {
+                Directory.CreateDirectory(outputFileDirectory);
+            }
+            
+#if NET461
+            var generatedCode = AppDomainCodeGenerator.GenerateCode(this.options);
+#else
+            var generatedCode = this.GenerateCodeInternal();
```

Comment:
it's a little odd how the conditional compilation was made, but it's correct. Ideally I would have the branching (but similar) code all in 1 file (or 2, but of the same shape). No big deal for this PR though

Reply:
you're right - it's definitely a bit odd. I'll address it, thanks